drivers/gpio/gpio_max149x6.h

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2024 Analog Devices Inc.
+ * Copyright (c) 2024 Baylibre SAS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define MAX149x6_READ 0
+#define MAX149x6_WRITE 1
+
+#define PRINT_ERR_BIT(bit1, bit2) if ((bit1) & (bit2)) LOG_ERR("[%s] %d", #bit1, bit1)
+#define PRINT_ERR(bit) if (bit) LOG_ERR("[DIAG] [%s] %d\n", #bit, bit)
+#define PRINT_INF(bit) LOG_INFO("[%s] %d\n", #bit, bit)
+#define LOG_DIAG(...)    Z_LOG(LOG_LEVEL_ERR, __VA_ARGS__)
+
+/**
+ * @brief Compute the CRC5 value for an array of bytes when writing to MAX149X6
+ * @param data - array of data to encode
+ * @param encode - action to be performed - true(encode), false(decode)
+ * @return the resulted CRC5
+ */
+static uint8_t max149x6_crc(uint8_t *data, bool encode)
+{
+	uint8_t crc5_start = 0x1f;
+	uint8_t crc5_poly = 0x15;
+	uint8_t crc5_result = crc5_start;
+	uint8_t extra_byte = 0x00;
+	uint8_t data_bit;
+	uint8_t result_bit;
+	int i;
+
+	/*
+	 * This is a custom implementation of a CRC5 algorithm, detailed here:
+	 * https://www.analog.com/en/app-notes/how-to-program-the-max14906-quadchannel-
+	 *					industrial-digital-output-digital-input.html
+	 */
+
+	for (i = (encode) ? 0 : 2; i < 8; i++) {
+		data_bit = (data[0] >> (7 - i)) & 0x01;
+		result_bit = (crc5_result & 0x10) >> 4;
+		if (data_bit ^ result_bit) {
+			crc5_result = crc5_poly ^ ((crc5_result << 1) & 0x1f);
+		} else {
+			crc5_result = (crc5_result << 1) & 0x1f;
+		}
+	}
+
+	for (i = 0; i < 8; i++) {
+		data_bit = (data[1] >> (7 - i)) & 0x01;
+		result_bit = (crc5_result & 0x10) >> 4;
+		if (data_bit ^ result_bit) {
+			crc5_result = crc5_poly ^ ((crc5_result << 1) & 0x1f);
+		} else {
+			crc5_result = (crc5_result << 1) & 0x1f;
+		}
+	}
+
+	for (i = 0; i < 3; i++) {
+		data_bit = (extra_byte >> (7 - i)) & 0x01;
+		result_bit = (crc5_result & 0x10) >> 4;
+		if (data_bit ^ result_bit) {
+			crc5_result = crc5_poly ^ ((crc5_result << 1) & 0x1f);
+		} else {
+			crc5_result = (crc5_result << 1) & 0x1f;
+		}
+	}
+
+	return crc5_result;
+}
+
+/*
+ * @brief Register write function for MAX14906
+ *
+ * @param dev - MAX14906 device config.
+ * @param addr - Register value to which data is written.
+ * @param val - Value which is to be written to requested register.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int max149x6_reg_transceive(const struct device *dev, uint8_t addr, uint8_t val,
```

Comment:
Why is this implemented in the header file?

Reply:
because it is shared from MAX14906 and MAX14916